Avoid NPE in WebAuthnAuthenticator on unknown user

thomasdarimont opened the pull request:
We now check if the user returned by the userId lookup exists. If the user does not exist, we show an `WEBAUTHN_ERROR_USER_NOT_FOUND` error in the form.

Fixes #29586

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

thomasdarimont committed:
Avoid NPE in WebAuthnAuthenticator on unknown user (#29586)

We now check if the user returned by the userId lookup exists.
If the user does not exist, we show an `WEBAUTHN_ERROR_USER_NOT_FOUND` error in the form.

Fixes #29586

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

jonkoops reviewed (dismissed):
LGTM,  thanks!

ahus1 commented:
@thomasdarimont - the tests fail: 

> Error: OR]   Run 3: WebAuthnErrorTest.errorPageIfUserDeletedDuringAuthentication:129 
> -- Expected: is "Unknown user authenticated by the Passkey."
> --      but: was "Failed to authenticate by the Passkey."

I'd say the message "Failed to authenticate by the Passkey." which is presented is preferred, as we want to avoid showing people that a user doesn't exist. It would be considered leaking information if a user exists in the system.

ahus1 reviewed (dismissed):
@thomasdarimont - the tests fail: 

> Error: OR]   Run 3: WebAuthnErrorTest.errorPageIfUserDeletedDuringAuthentication:129 
> -- Expected: is "Unknown user authenticated by the Passkey."
> --      but: was "Failed to authenticate by the Passkey."

I'd say the message "Failed to authenticate by the Passkey." which is presented is preferred, as we want to avoid showing people that a user doesn't exist. It would be considered leaking information if a user exists in the system.

jonkoops commented:
>  It would be considered leaking information if a user exists in the system.

This requires that someone uses a key they own, no? So how would that leak information? We also log similar message in some code just above it.

ahus1 commented:
> This requires that someone uses a key they own, no? So how would that leak information? We also log similar message in some code just above it.

@jonkoops - this is not logging some information, but showing them in the browser AFAIK. 

Maybe @thomasdarimont can clarify when this is shown.

thomasdarimont commented:
The proposed "Unknown user authenticated by the Passkey." error message can be triggered, if a user registered a passkey before and this passkey is stored in a password manager (e.g. google password manager). When the `PasskeysConditionalUIAuthenticator` is used the user can select a previously registered passkey from a list provided by the browser / password manager integration. If the user selects the passkey and proceeds with the authentication, a `userHandle` with the user uuid of that user is sent to Keycloak alongside other metadata.

If the user is removed inbetween (user storage federation update, account expiration, or other cleanup) the user with that ID no longer exists in the Keycloak database, hence the lookup by the `userHandle` will fail and return `null`.

The message "Unknown user authenticated by the Passkey."  expresses exactly that (error) situation. -> a user attempted to use a valid passkey bound to a user that no longer exists.

jonkoops commented:
> this is not logging some information, but showing them in the browser AFAIK.

Yes, that is what I meant. Used the wrong terminology. Personally I think the "Unknown user authenticated by the Passkey." is fine, but the tests should of course be passing.

thomasdarimont commented:
I use the same error message as is used, when the authenticator credential isn't valid, with slightly different details in the event.
See: https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/authentication/authenticators/browser/WebAuthnAuthenticator.java#L235

> ... but the tests should of course be passing.

That's my current struggle... I cannot execute the test on my machine :-/

ahus1 commented:
@thomasdarimont 

> The proposed "Unknown user authenticated by the Passkey." error message can be triggered, if a user registered a passkey before and this passkey is stored in a password manager (e.g. google password manager).

Thanks for the explanation. Still I wonder if a user enumeration would be possible using this: I would register a passkey to a web site with a random username, by for example faking the TLS and everything that comes with it to allow Google to register the passkey. Then I go to the real website and it tells me that the user doesn't exist. 

So IMHO I think user enumeration is possible, as you can't trust the wallet with the Webauthn keys. Therefore, the error message shouldn't reveal if the user doesn't exist or the passkey is wrong.

thomasdarimont commented:
Currently an attacker with a valid bit unknown WebAuthN credentials / Passkeys attempts a login with the Passkeys, the can use the NPE as an indicator for whether a user is present. Of they get am NPE the user with the userHandle (user ID) does not exist.
If the user does exist the supplied Passkeys credential is check if it is valid.
If it is invalid the error message from above is shown.

I think if we use the same error message if a user does not exist an attacker should not be able to tell if the user exists or not based on the message.

However, it is probably still possible to use a timing attack in this case.

ahus1 commented:
> if we use the same error message if a user does not exist an attacker should not be able to tell if the user exists or not based on the message.

I would be ok with that approach.

thomasdarimont committed:
Tests for avoid NPE in WebAuthnAuthenticator on unknown user (#29586)

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

thomasdarimont commented:
Investigated the failing test a bit and found the cause... the test `org.keycloak.testsuite.webauthn.account.WebAuthnErrorTest#errorPageWithTimeout` I copied it from seems to be wrong too :)
<img width="1496" alt="image" src="https://github.com/user-attachments/assets/783b418e-a440-4826-bafc-433a6f20eeab" />

The test "passes" but just because it produces the same error message `NotAllowedError: The operation either timed out or was not allowed. See: https://www.w3.org/TR/webauthn-2/#sctn-privacy-considerations-client.` is produced which yields the `Failed to authenticate by the Passkey.` error message. in the end.

If I skip over this issue (by setting `errorMsgFromWebAuthnApi=null` in the debugger in  [WebAuthnAuthenticator]( https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/authentication/authenticators/browser/WebAuthnAuthenticator.java#L139)) I see the same error message as I was using in my test "Unknown user authenticated by the Passkey.".

I now use the `org.keycloak.testsuite.webauthn.WebAuthnRegisterAndLoginTest#webAuthnPasswordlessAlternativeWithWebAuthnAndPassword` as reference.

thomasdarimont commented:
@ahus1 is this PR now okay for you?

thomasdarimont commented:
Thanks for the hint. I updated the test to use a dedicated test user, apprently the user I used (and deleted in the test) is used by the other rests and there is no environment reset after/before a test runs.

ahus1 reviewed (commented):
I gave it a try and it works now as expected that an error message appears when the user is deleted while the authentication flow is in progress. It also fixes the case when you have only the passwordless authentication where you pick the username on your device and the user has already been removed from the system. 

I found that there is a slightly different behavior when the user doesn't exist, or the passkey has been removed from the user. 

In the same method below, there is the same error response, but the flow is cancelled when the user exists but the passkey has been removed.

https://github.com/keycloak/keycloak/pull/37427/files#diff-6dc319037c4a9a0d86c5a36f26403222ded602c6257ed318b8c3e0e97774623bR238-R244

Due to the cancelling, the error message is never shown. And there is a different behavior for cases when the user has been deleted, or when the passkey has been deleted. The observed behavior from an attacker should be identical. 
Also it is not helpful to a regular user that the login flow is cancelled, as they won't be able to see the error message. 

So I'd say the 'context.cancelLogin()` should be removed in line 243. This could happen in this PR, or in a new issue and pull request.

thomasdarimont commented:
@ahus1 Thanks for taking a look and the additional feedback.

instead of just removing the line with `context.cancelLogin();`, how about replacing it with `context.attempted();` to at least document that the authenticator was tried?

thomasdarimont committed:
Avoid possibility of user enumeration in WebAuthnAuthenticator

Fixes #29586

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

ahus1 commented:
> Instead of just removing the line with `context.cancelLogin();`, how about replacing it with `context.attempted();` to at least document that the authenticator was tried?

Hi Thomas - I have to admit I'm not that deep into the authentication flow. My only comment here is to make it behave the same if the user doesn't exist, and if the authentication key hasn't been found. I also prefer to display the error message in the Keycloak UI instead of redirecting to the application directly, so people can correct their inputs (different user name, different selected key).

I assume with that input you can make a better choice than me if `attempted()` makes sense. I only say that IMHO `cancelled()` doesn't make sense.

ahus1 commented:
@thomasdarimont - I just looked at the docs of "attempted": 

> There was no failure or challenge.  The authenticator was attempted, but not fulfilled.  If the current execution requirement is alternative or optional, then this status is ignored by the flow.

This doesn't sound right, and I am leaning towards merging it as-is. Please let me know if you agree.
